### PR TITLE
Adds a sharing button to the preview mode of the editor.

### LINKED ISSUE #3792: Sharing: Would be nice to have a way to share a link to a post from posts in My Sites
We've never had a way to share a link from the posts or pages feature, and it would be nice to have.  The best place might be the "detail" screen that precedes the active editor.

### PR DESCRIPTION
**Pull request by diegoreymendez**
Fixes #3792.

To test:
1. Launch the app
2. Go to the list of blogs for a site.
3. Tap on an existing post.
4. While in preview mode, tap the new share button.

Needs review: @aerych

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Post/PostSharingController.swift [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        controller.modalPresentationStyle = .Popover
+        viewController.presentViewController(controller, animated: true, completion: nil)
+        if let presentationController = controller.popoverPresentationController {
+            presentationController.permittedArrowDirections = .Unknown
+            presentationController.sourceView = anchorView
+            presentationController.sourceRect = anchorView.bounds
+        }
+    }
+    
+    func sharePost(post: Post, fromView anchorView:UIView, inViewController viewController:UIViewController) {
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/PostSharingController.swift]: It might be nice to extend `PostContentProvider` to include methods for retrieving tags and a permalink, not necessarily by those names. Then there could be a single `sharePost` method accepting either Post or ReaderPost (or a Page for that matter).  
Just a thought. Otherwise what is here is fine.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 2]
```diff
@@ -243,6 +247,8 @@ - (instancetype)initWithPost:(AbstractPost *)post
             
             _ownsPost = YES;
         }
+        
+        _sharingController = [[PostSharingController alloc] init];
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Probably no need to create this during in it since its not likely to be used very often.  There's probably no need to save a reference in a property either.  Maybe just create / recreate on demand.

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 3]
```diff
@@ -48,6 +48,8 @@ import WordPressComAnalytics
     private var needsRefreshCachedCellHeightsBeforeLayout = false
     private var didSetupView = false
     private var listentingForBlockedSiteNotification = false
+    
+    private let sharingController = PostSharingController()
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: Probably no need to keep a reference.  Just create on demand.

File: WordPress/Classes/ViewRelated/Reader/ReaderPostMenu.swift [hunk 4]
```diff
@@ -4,6 +4,8 @@ import SVProgressHUD
 public class ReaderPostMenu
 {
     public static let BlockSiteNotification = "ReaderPostMenuBlockSiteNotification"
+    
+    private static let sharingController = PostSharingController()
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderPostMenu.swift]: Probably no need to keep a reference.  Just create on demand.

**Comment by aerych** [comment]
Hiya @diegoreymendez :)

I left a few comments in code. I also took the branch for a test drive and noticed a few things. 

Looks like there's some unhappy constraints on the iPad.  I get the following in the console after tapping the share button:
```
2016-04-04 13:46:53.427 WordPress[11218:165035] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x7c857470 V:[_UIDimmingKnockoutBackdropView:0x7c8583c0(>=44)]>",
    "<NSLayoutConstraint:0x7b102e00 UIView:0x7b1dfe90.height == UIView:0x7b1dfd40.height>",
    "<NSLayoutConstraint:0x7b628ab0 _UIDimmingKnockoutBackdropView:0x7c8583c0.top == UIView:0x7b1dfe90.top>",
    "<NSLayoutConstraint:0x7b1000f0 _UIAlertControllerView:0x7b1e0a40.height == UIView:0x7b1dfd40.height>",
    "<NSLayoutConstraint:0x7c886dc0 _UIAlertControllerView:0x7b1e0a40.height == UIView:0x7b093210.height>",
    "<NSLayoutConstraint:0x7c88b710 _UIDimmingKnockoutBackdropView:0x7c8583c0.bottom == UICollectionView:0x7b95a800.bottom>",
    "<NSLayoutConstraint:0x7c888530 UIView:0x7b1dfe90.bottom == UICollectionView:0x7b95a800.bottom>",
    "<NSLayoutConstraint:0x7c83f630 'UIView-Encapsulated-Layout-Height' V:[UIView:0x7b093210(21)]>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x7c888530 UIView:0x7b1dfe90.bottom == UICollectionView:0x7b95a800.bottom>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
2016-04-04 13:46:53.428 WordPress[11218:165035] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x7b106170 V:[UIView:0x7b1dfe90(>=44)]>",
    "<NSLayoutConstraint:0x7b102e00 UIView:0x7b1dfe90.height == UIView:0x7b1dfd40.height>",
    "<NSLayoutConstraint:0x7b1000f0 _UIAlertControllerView:0x7b1e0a40.height == UIView:0x7b1dfd40.height>",
    "<NSLayoutConstraint:0x7c886dc0 _UIAlertControllerView:0x7b1e0a40.height == UIView:0x7b093210.height>",
    "<NSLayoutConstraint:0x7c83f630 'UIView-Encapsulated-Layout-Height' V:[UIView:0x7b093210(21)]>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x7b106170 V:[UIView:0x7b1dfe90(>=44)]>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
2016-04-04 13:46:53.436 WordPress[11218:165035] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x7b1062b0 V:[UIView:0x7b1dffe0(>=44)]>",
    "<NSLayoutConstraint:0x7b1061a0 V:|-(0)-[UIView:0x7b1dffe0]   (Names: '|':UIView:0x7b1dfe90 )>",
    "<NSLayoutConstraint:0x7b106230 V:[UICollectionView:0x7b95a800]-(0)-|   (Names: '|':UIView:0x7b1dffe0 )>",
    "<NSLayoutConstraint:0x7b102e00 UIView:0x7b1dfe90.height == UIView:0x7b1dfd40.height>",
    "<NSLayoutConstraint:0x7b1000f0 _UIAlertControllerView:0x7b1e0a40.height == UIView:0x7b1dfd40.height>",
    "<NSLayoutConstraint:0x7b7e56b0 _UIAlertControllerView:0x7b1e0a40.height == UIView:0x7b093210.height>",
    "<NSLayoutConstraint:0x7b739cd0 UIView:0x7b1dfe90.bottom == UICollectionView:0x7b95a800.bottom>",
    "<NSLayoutConstraint:0x7c83f630 'UIView-Encapsulated-Layout-Height' V:[UIView:0x7b093210(21)]>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x7b1062b0 V:[UIView:0x7b1dffe0(>=44)]>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.

```

And the popover is broken:
simulator screen shot apr 4 2016 1 45 17 pm


Lastly, it looks like the Share button stays in its pressed state after being tapped, even after the actionsheet / popover is dismissed.

**Comment by diegoreymendez** [comment]
Ty @aerych, I'll take it from here.

File: WordPress/Classes/ViewRelated/Post/PostSharingController.swift (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostSharingController.swift]: I started aiming for that, but found a few technical issues making it a bit harder than I would have liked for the scope of this PR.  More specifically: the prototype didn't like having methods named as some of the properties from the classes implementing the protocol.

Regardless of that, though, I was hoping to review this particular issue separately because I feel this could have a better solution.  If not, I'll make sure to submit a separate PR with a protocol to take care of this (but I'd still like to abstract that "issue" from this specific PR).

Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/PostSharingController.swift]: No worries. I'd taken a quick peek at the code to see what an implementation looked like and I could see where it might not be straight forward.  I was hoping I was mistaken. ;)  Let's keep both methods for now.

**Comment by diegoreymendez** [comment]
@aerych - Just fixed the issue.  I now have a much smaller issue that I can't find a good fix for.

To repro:
1. Launch the app in an iPad.
2. Go to the list of posts for a blog.
3. Preview the post.
4. Tap share.
5. The share popup should come up correctly now.
6. Tap outside of the popup to make it go away.

Expected: the share button should go back to white.
Current behavior: the share button never goes back to white.

The bug is pretty minor so I'd like to push this forward anyway.  My first question is if you happen to know what may be causing the issue I'm seeing (since you've worked on sharing in the past).  If not, let me know and I'll open a separate issue for this.

**Comment by aerych** [comment]
> The bug is pretty minor so I'd like to push this forward anyway

I wouldn't say its a minor bug since it looks pretty broken when it happens. It occurs on the iPhone as well as iPad. 

It looks like adding a `touchUpInside` event to `WPButtonForNavigationBar` to reset the alpha would fix it.

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 5]
```diff
@@ -558,6 +556,14 @@ import WordPressComAnalytics
             presentViewController(alertController, animated: true, completion: nil)
         }
     }
+    
+    private func sharePost(postID: NSManagedObjectID, fromView anchorView: UIView) {
+        if let post = self.postWithObjectID(postID) {
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: Nitpicky, I know, but maybe use a `guard` instead and exit early vs doing the work inside the block.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 6]
```diff
@@ -781,6 +782,19 @@ - (void)showFailedMediaBeforeEditAlert
     [self presentViewController:alertController animated:YES completion:nil];
 }
 
+#pragma mark - Toolbar options
+
+- (void)sharePost
+{
+    if ([self.post isKindOfClass:[Post class]]) {
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Again, super nitpicky,  but maybe return if its not a post class and move the work out of the conditional?

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift (hunk 5 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: Any particular reason why the change would be necessary other than preference of one vs the other?

`if let` is a pretty standard way of avoiding `nil` values.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 6 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Same as above for this.  Early returns are just fine for longer methods, but don't really see them as positive for smaller methods such as this one (the method ends up having an extra line of code for no gain).

**Comment by diegoreymendez** [comment]
> I wouldn't say its a minor bug since it looks pretty broken when it happens. It occurs on the iPhone as well as iPad.

Minor as in non-breaking.  Even though it's kind of ugly it's only a visual glitch, as opposed to a showstopper.  I spent all of yesterday trying to find the culprit but couldn't, and it feels like this bug may be on Apple's side at this point.

I don't think sending the button an event is a good idea because it could bring out other issues.  I'd rather close this PR without merging in that case, but it'd be a pity to not add a feature that works because of a visual glitch that's outside of our hands to fix.

I'll test some more and close this PR if I can't find a good solution.

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m (hunk 6 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Yah it was nitpicky :)  Either way is fine.

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift (hunk 5 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: Seems like with swift there is a preference for guard for semantic reasons but the end result is the same. [One opinion](https://www.natashatherobot.com/swift-guard-better-than-if/).  Since the work should never be done if the assignment results in nil, it would be "swiftier" to use guard and exit. 
Again, this was a nitpick so its up to you. :)

**Comment by aerych** [comment]
Summarizing a quick chat: 
The sharing button is an instance of `WPButtonForNavigationBar`, which is already using event handlers to update its alpha in response to touch and drag events. My point was that adding a new event to reset the alpha on touchUpInside would reset the alpha state correctly. (Confirmed by testing before I left my original comment.) 
However, other approaches that might make sense include removing the existing animations that reset the button alpha, or to just use a plain ol' `UIBarButtonItem`.

**Comment by diegoreymendez** [comment]
I added a touchUp event instead of just removing `WPButtonForNavigationBar`.  The reason is that the class has support for custom spacing - so removing it would mean also taking care of addressing the custom spacing support, which sounded like a bit too far from the scope of this PR.

**Comment by aerych** [comment]
Retested all the things and it looks good :) 
:shipit:

**Comment by diegoreymendez** [comment]
@aerych - Fixed.  An oddity with `UIActivityViewController`, it seems!

**Comment by aerych** [comment]
Yay!!
:shipit: as soon as the merge conflict is resolved :)

**Comment by diegoreymendez** [comment]
Ty Eric!
